fastapi/routing.py

```diff
@@ -52,6 +52,8 @@ def serialize_response(
             errors.extend(errors_)
         if errors:
             raise ValidationError(errors)
+        if skip_defaults and isinstance(response, BaseModel):
+            exclude |= value.__fields_set__.difference(response.__fields_set__)
         return jsonable_encoder(
```

Comment:
```python
        if isinstance(response, BaseModel):
            value = response.dict(skip_defaults=skip_defaults)
```
might also work, but I'm not sure if that would play differently with include/exclude/`jsonable_encoder`.